File: resources/[vrp]/vrp-lib/client/cl_ui.lua
local registered = {}

function RegisterUICallback(name, cb)
  local function interceptCb(data, innerCb)
    cb(data, function(result)
      if result.meta.ok then
        result.meta.message = "done"
      end
      innerCb(result)
    end)
  end
  AddEventHandler(('_vpx_uiReq:%s'):format(name), interceptCb)

  if (GetResourceState("vrp-ui") == "started") then exports["vrp-ui"]:RegisterUIEvent(name) end

  registered[#registered + 1] = name
end

function SendUIMessage(data)
  exports["vrp-ui"]:SendUIMessage(data)
end

function SetUIFocus(hasFocus, hasCursor)
  exports["vrp-ui"]:SetUIFocus(hasFocus, hasCursor)
end

function GetUIFocus()
  return exports["vrp-ui"]:GetUIFocus()
end

AddEventHandler("_vpx_uiReady", function()
  for _, eventName in ipairs(registered) do
    exports["vrp-ui"]:RegisterUIEvent(eventName)
  end
end)

File: resources/[vrp]/vrp-lib/client/cl_ui.js
const registered = [];

function RegisterUICallback(name, cb) {
    AddEventHandler(`_vpx_uiReq:${name}`, cb);

    if (GetResourceState('vrp-ui') === 'started') exports['vrp-ui'].RegisterUIEvent(name);

    registered.push(name);
}

function SendUIMessage(data) {
    exports['vrp-ui'].SendUIMessage(data);
}

function SetUIFocus(hasFocus, hasCursor) {
    exports['vrp-ui'].SetUIFocus(hasFocus, hasCursor);
}

function GetUIFocus() {
    return exports['vrp-ui'].GetUIFocus();
}

AddEventHandler('_vpx_uiReady', () => {
    registered.forEach((eventName) => exports['vrp-ui'].RegisterUIEvent(eventName));
});